src/web/source.rs

```diff
@@ -222,34 +222,64 @@ pub(crate) async fn source_browser_handler(
         }
     };
 
-    let archive_storage = spawn_blocking({
+    let (archive_storage, latest_build_id): (bool, Option<i32>) = spawn_blocking({
+        // we probably could not use a SQL statement here, but rather a more selective
+        // version of `CrateDetails`, which doesn't fetch everything.
         let pool = pool.clone();
         let name = name.clone();
         let version = version.clone();
         move || {
             let mut conn = pool.get()?;
-            Ok(conn
-                .query_one(
-                    "SELECT archive_storage
-                     FROM releases
-                     INNER JOIN crates ON releases.crate_id = crates.id
-                     WHERE
-                         name = $1 AND
-                         version = $2",
-                    &[&name, &version],
-                )?
-                .get::<_, bool>(0))
+            let row = conn.query_one(
+                "SELECT
+                    releases.archive_storage,
+                    (
+                        SELECT id
+                        FROM builds
+                        WHERE
+                            builds.rid = releases.id AND
+                            builds.build_status = TRUE
+                        ORDER BY build_time DESC
+                        LIMIT 1
+                    ) AS latest_build_id
+                 FROM releases
+                 INNER JOIN crates ON releases.crate_id = crates.id
+                 WHERE
+                     name = $1 AND
+                     version = $2",
+                &[&name, &version],
+            )?;
+            Ok((row.get(0), row.get(1)))
         }
     })
     .await?;
 
     // try to get actual file first
     // skip if request is a directory
-    let blob = if !path.ends_with('/') {
-        storage
-            .fetch_source_file(&name, &version, &path, archive_storage)
+    let blob = if !path.ends_with('/')
+        && storage
+            .source_file_exists(
```

Comment:
Instead of pre-checking here, can we just attempt to fetch and then map a `PathNotFoundError` to `None`.

Reply:
generally the existance check would be relatively fast because it's just using the local archive index, but this is still a valid point, and I updated the code.